src/conversions/chrono.rs

```diff
@@ -39,143 +38,220 @@
+        {
+            let date: &PyDate = ob.downcast()?;
+            py_date_to_naive_date(date)
+        }
+        #[cfg(Py_LIMITED_API)]
+        {
+            check_type(ob, date_class(ob.py()), "PyDate")?;
+            py_date_to_naive_date(ob)
+        }
     }
 }
 
 impl ToPyObject for NaiveTime {
     fn to_object(&self, py: Python<'_>) -> PyObject {
-        (*self).into_py(py)
-    }
-}
-
-impl IntoPy<PyObject> for NaiveTime {
-    fn into_py(self, py: Python<'_>) -> PyObject {
         let TimeArgs {
             hour,
             min,
             sec,
             micro,
             truncated_leap_second,
         } = self.into();
+        #[cfg(not(Py_LIMITED_API))]
         let time = PyTime::new(py, hour, min, sec, micro, None).expect("Failed to construct time");
+        #[cfg(Py_LIMITED_API)]
+        let time = time_class(py)
+            .call1((hour, min, sec, micro))
+            .expect("failed to construct datetime.time");
         if truncated_leap_second {
             warn_truncated_leap_second(time);
         }
         time.into()
     }
 }
 
+impl IntoPy<PyObject> for NaiveTime {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        self.to_object(py)
+    }
+}
+
 impl FromPyObject<'_> for NaiveTime {
     fn extract(ob: &PyAny) -> PyResult<NaiveTime> {
-        let time: &PyTime = ob.downcast()?;
-        py_time_to_naive_time(time)
+        #[cfg(not(Py_LIMITED_API))]
+        {
+            let time: &PyTime = ob.downcast()?;
+            py_time_to_naive_time(time)
+        }
+        #[cfg(Py_LIMITED_API)]
+        {
+            check_type(ob, time_class(ob.py()), "PyTime")?;
+            py_time_to_naive_time(ob)
+        }
     }
 }
 
 impl ToPyObject for NaiveDateTime {
     fn to_object(&self, py: Python<'_>) -> PyObject {
         naive_datetime_to_py_datetime(py, self, None)
-            .expect("failed to construct datetime")
-            .into()
     }
 }
 
 impl IntoPy<PyObject> for NaiveDateTime {
     fn into_py(self, py: Python<'_>) -> PyObject {
-        ToPyObject::to_object(&self, py)
+        self.to_object(py)
     }
 }
 
 impl FromPyObject<'_> for NaiveDateTime {
-    fn extract(ob: &PyAny) -> PyResult<NaiveDateTime> {
-        let dt: &PyDateTime = ob.downcast()?;
+    fn extract(dt: &PyAny) -> PyResult<NaiveDateTime> {
+        #[cfg(not(Py_LIMITED_API))]
+        let dt: &PyDateTime = dt.downcast()?;
+        #[cfg(Py_LIMITED_API)]
+        check_type(dt, datetime_class(dt.py()), "PyDateTime")?;
+
         // If the user tries to convert a timezone aware datetime into a naive one,
         // we return a hard error. We could silently remove tzinfo, or assume local timezone
         // and do a conversion, but better leave this decision to the user of the library.
+        #[cfg(not(Py_LIMITED_API))]
         if dt.get_tzinfo().is_some() {
             return Err(PyTypeError::new_err("expected a datetime without tzinfo"));
         }
+        #[cfg(Py_LIMITED_API)]
```

Comment:
Maybe factor this out into a binding, e.g. `let has_tzinfo = ...;` to avoid duplicating the error message construction.